src/coreclr/jit/morph.cpp

```diff
@@ -12478,6 +12484,65 @@ GenTree* Compiler::fgOptimizeEqualityComparisonWithConst(GenTreeOp* cmp)
             op1->SetVNsFromNode(cmp);
 
             DEBUG_DESTROY_NODE(cmp);
+
+            if (op1->OperIs(GT_GE, GT_LE) && (op1->gtGetOp2()->IsIntegralConst() || op1->gtGetOp1()->IsIntegralConst()))
+            {
+                GenTree* conNode = op1->gtGetOp2()->IsIntegralConst() ? op1->gtGetOp2() : op1->gtGetOp1();
+
+                IntegralRange valRange = IntegralRange::ForNode(conNode, this);
+
+#if defined(HOST_X86) || defined(HOST_ARM)
+                ssize_t valMin =
+                    (int32_t)IntegralRange::SymbolicToRealValue(valRange.LowerBoundForType(conNode->TypeGet()));
+                ssize_t valMax =
+                    (int32_t)IntegralRange::SymbolicToRealValue(valRange.UpperBoundForType(conNode->TypeGet()));
+#else
+                ssize_t valMin = IntegralRange::SymbolicToRealValue(valRange.LowerBoundForType(conNode->TypeGet()));
+                ssize_t valMax = IntegralRange::SymbolicToRealValue(valRange.UpperBoundForType(conNode->TypeGet()));
+#endif
+                // Folds
+                // 1. byte x <= 0 => true
+                // 2. int  x <= int.MinValue => true
+                // 3. long x <= long.MaxValue => true
+                // 4. byte x >= byte.MaxValue => true
+                // 5. int  x >= int.MaxValue => true
+                // 6. long x >= long.MaxValue => true
+                //
+                // when const is RHS:
+                if (op1->gtGetOp2()->IsIntegralConst())
+                {
+
+                    if ((op1->OperIs(GT_GE) && conNode->IsIntegralConst(valMin)) ||
+                        (op1->OperIs(GT_LE) && conNode->IsIntegralConst(valMax)))
+                    {
+                        GenTree* ret = gtNewIconNode(1, TYP_INT);
+                        ret->SetVNsFromNode(op1);
+                        DEBUG_DESTROY_NODE(op1);
+                        return fgMorphTree(ret);
+                    }
+                } // when const is LHS:
+                else
+                {
+                    if ((op1->OperIs(GT_GE) && conNode->IsIntegralConst(valMax)) ||
+                        (op1->OperIs(GT_LE) && conNode->IsIntegralConst(valMin)))
+                    {
+                        GenTree* cmpSideEffects = nullptr;
+                        GenTree* ret            = gtNewIconNode(1, TYP_INT);
+
+                        gtExtractSideEffList(op1, &cmpSideEffects);
```

Comment:
Something simple like:
```csharp
class C { public byte B; }
bool TestFold(C c) => c.B <= 255;
TestFold(null);
```
ought to work. If you folded this away completely it would not throw `NullReferenceException` anymore.

Reply:
Also, it would be nice to add a test case with these cases and @stephentoub's cases from #70145.

Reply:
I would love to do this. I have never written any tests for Jit, so I need your guidance. I don't even know which directory to put them (after src\tests\JIT\)  :)

Reply:
I would add it in `src\tests\JIT\opt\Compares\compares.cs/csproj`. You can copy the template from `src\tests\JIT\opt\OptimizeBools\optboolsreturn.cs/csproj`.

Reply:
> Something simple like:
> 
> ```cs
> class C { public byte B; }
> bool TestFold(C c) => c.B <= 255;
> TestFold(null);
> ```

Unfortunately the exception is being thrown at some earlier step, is there another way to reproduce the side effects?

Reply:
Yeah, you are right. I face an issue. The extracted side effects node is morphed, so I can't morph the new node created by `gtNewOperNode(GT_COMMA, TYP_INT, cmpSideEffects, result);`. So I'm considering to skip the optimization if I there is any side effects. Is it all right?

Reply:
I don't think you should need to call morph again here. You can check how the `NI_System_Runtime_CompilerServices_RuntimeHelpers_IsKnownConstant` is handled in post-order, it has the same problem as you do where it needs to extract side effects. You might need to set `GTF_DEBUG_NODE_MORPHED` manually on the newly created comma.

Reply:
I had tried to follow your recomendations up and It seemed like working well, but then I did SPMI replay and it highlighted an assertion failure. I'm stuck with it and can't figure out why it keeps showing up. The issue is:
`Assertion failed 'OperIs(GT_NOP, GT_CALL, GT_COMMA) || OperIsCompare() || OperIsLong() || OperIsSimdOrHWintrinsic() || IsCnsVec()' in 'BlockingConfig:.cctor()' during 'Lowering nodeinfo' (IL size 251; hash 0x0bf431ba; Tier0)`
It goes away if I exclude TYP_SHORT from the optimization. Is there any way to force JitDump or anything helpfull for debugging during spmi replays? 
I would love to learn how you debug such replay failures and figure out what is wrong with the new code :)

Reply:
The SPMI replay logs output to help with that. So for instance you should see something like:
```scala
[20:37:41] Running SuperPMI replay of C:\dev\dotnet\spmi\mch\5868685e-b877-4ef5-83f0-73d601e50013.Linux.arm\libraries.crossgen2.Linux.arm.checked.mch
[20:37:49] Compilation failures
[20:37:51]
[20:37:51] Repro 3305 .mc file(s) created for failures:
[20:37:51] C:\dev\dotnet\spmi\repro.libraries.crossgen2.Linux.arm.checked\repro-100567.mc
[20:37:51] C:\dev\dotnet\spmi\repro.libraries.crossgen2.Linux.arm.checked\repro-104855.mc
[20:37:51] C:\dev\dotnet\spmi\repro.libraries.crossgen2.Linux.arm.checked\repro-105267.mc
...
[20:37:52]
[20:37:52] To run a specific failure (replace JIT path and .mc filename as needed):
[20:37:52]
[20:37:52] C:\dev\dotnet\runtime3\artifacts\tests\coreclr\windows.x86.Checked\Tests\Core_Root\superpmi.exe -target arm C:\dev\dotnet\runtime3\artifacts\tests\coreclr\windows.x86.Checked\Tests\Core_Root\clrjit_universal_arm_x86.dll C:\dev\dotnet\spmi\repro.libraries.crossgen2.Linux.arm.checked\xxxxx.mc
```
(from a recent run of mine, the exact details will be different for you). Then my typical approach is to use the smallest repro file here. The repro files contain just a single function being compiled, so you can just set `COMPlus_JitDump=*` (or `COMPlus_NgenDump=*` if the context is from a crossgenned context) before running superpmi on the repro file.